### PR TITLE
Enrich etcd server cert SANS to prevent DNS lookups with no result

### PR DESCRIPTION
**Pull request by rfranzke**
**What this PR does / why we need it**:
We now re-create the server certificate for all etcds of all shoots and enrich the SANS.
This is to prevent undesired NXDOMAIN DNS results when the kube-apiserver talks to etcd.
Previously, we always saw a lot of `NXDOMAIN` results, e.g.

```
2019-02-22T07:45:24.493Z [INFO] 10.223.128.26:44597 - 57208 "A IN etcd-main-client.shoot--dev--test.svc.shoot--dev--test.svc.cluster.local. udp 102 false 512" NXDOMAIN qr,aa,rd,ra 195 0.000201666s
2019-02-22T07:45:24.493Z [INFO] 10.223.128.26:45438 - 11118 "AAAA IN etcd-main-client.shoot--dev--test.svc.shoot--dev--test.svc.cluster.local. udp 102 false 512" NXDOMAIN qr,aa,rd,ra 195 0.000237213s
2019-02-22T07:45:24.494Z [INFO] 10.223.128.26:46011 - 43203 "AAAA IN etcd-main-client.shoot--dev--test.svc.svc.cluster.local. udp 79 false 512" NXDOMAIN qr,aa,rd,ra 172 0.000250539s
2019-02-22T07:45:24.494Z [INFO] 10.223.128.26:52359 - 57505 "A IN etcd-main-client.shoot--dev--test.svc.svc.cluster.local. udp 79 false 512" NXDOMAIN qr,aa,rd,ra 172 0.000241489s
2019-02-22T07:45:24.496Z [INFO] 10.223.128.26:56367 - 16269 "A IN etcd-main-client.shoot--dev--test.svc.cluster.local. udp 75 false 512" NOERROR qr,aa,rd,ra 148 0.000146696s
2019-02-22T07:45:24.496Z [INFO] 10.223.128.26:34198 - 57522 "AAAA IN etcd-main-client.shoot--dev--test.svc.cluster.local. udp 75 false 512" NOERROR qr,aa,rd,ra 168 0.000113945s
```

which put unneeded load onto the seeds DNS server. At scale this can lead to lookup times `> 1s` which can make the `/healthz` report of the kube-apiserver fail.

**Release note**:

```improvement operator
NONE
```

### CODE REVIEW

**Review by mvladev — COMMENTED** [review]
Isn't it better to check the actual secret's certificate and check if the DNS names are present?

I would rather use `cert-manager` for those secrets as it might be a better solution.

**Comment by rfranzke** [comment]
> Isn't it better to check the actual secret's certificate and check if the DNS names are present?

Hm yeah, in this case it's probably better. This implementation was influenced by thoughts of how future certificate rotation could look like (without changing the content of the actual secret like today).

> I would rather use `cert-manager` for those secrets as it might be a better solution.

Not sure, I don't want Gardener to rely on cert-manager for generating these certificates. It's easily possible to do it within the Golang code itself.

**Comment by timuthy** [comment]
Is there any definition of k8s that the first search path is and will be `<namespace>.svc.cluster.local` by default? 

Because this basically the fact we rely on here, don't we?

**Comment by rfranzke** [comment]
Updated to compare the actual SANS with the desired SANS. cc @mvladev 

@timuthy good question, I guess there is no contract, it's just convention that it's like this. I only found https://kubernetes.io/docs/tasks/administer-cluster/dns-debugging-resolution/

**Comment by rfranzke** [comment]
https://github.com/kubernetes/dns/blob/master/docs/specification.md FYI

File: pkg/operation/botanist/secrets.go [hunk 1]
```diff
@@ -760,27 +761,33 @@ func (b *Botanist) fetchExistingSecrets() (map[string]*corev1.Secret, error) {
...
+	val, err := secrets.LoadCertificate(certificateETCDServer, secret.Data[secrets.DataKeyPrivateKey], secret.Data[secrets.DataKeyCertificate])
+	if err != nil {
+		return err
+	}
+	certificate, ok := val.(*secrets.Certificate)
+	if !ok {
+		return fmt.Errorf("could not convert certificate into *secrets.Certificate struct")
+	}
+
+	if crt := certificate.Certificate; crt != nil && reflect.DeepEqual(crt.DNSNames, dnsNamesForEtcd(b.Shoot.SeedNamespace)) {
```
Comment (mvladev) [inline pkg/operation/botanist/secrets.go]: isn't it better to use `sets`  and `HasAll` from apimachinery for this?
`reflect.DeepEqual` will fail when comparing `["a", "b"]` and `["b", "a"]` (slice order)

**Review by mvladev — COMMENTED** [review]
small nit

File: pkg/operation/botanist/secrets.go (hunk 1 above)
Comment (rfranzke) [reply pkg/operation/botanist/secrets.go]: Sure, agreed. Please check again.

**Review by timuthy — APPROVED** [review]

**Review by mvladev — APPROVED** [review]
/lgtm

**Comment by vlerenc** [comment]
Wow, good catch. Must have been difficult to pinpoint.

**Pull request merged**